pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -1068,15 +1080,49 @@ impl Timeline {
             Ok((keys_above_horizon, img))
         }
 
-        let mut delta_layer_writer = DeltaLayerWriter::new(
-            self.conf,
-            self.timeline_id,
-            self.tenant_shard_id,
-            all_key_values.first().unwrap().0,
-            gc_cutoff..max_lsn, // TODO: off by one?
-            ctx,
-        )
-        .await?;
+        async fn flush_deltas(
+            deltas: &mut Vec<(Key, Lsn, crate::repository::Value)>,
+            last_key: Key,
+            delta_split_points: &[Key],
+            current_delta_split_point: &mut usize,
+            tline: &Arc<Timeline>,
+            gc_cutoff: Lsn,
+            ctx: &RequestContext,
+        ) -> anyhow::Result<Option<ResidentLayer>> {
+            // Check if we need to split the delta layer. We split at the original delta layer boundary to avoid
+            // overlapping layers.
+            let mut need_split = false;
```

Comment:
Hm, I don't see immediately how that overlap can happen. ASCII art or a _small_ test case that fails before this change can help.

Reply:
https://github.com/neondatabase/neon/pull/8078/commits/d3f7d5e8a7738cb035b62953cdd13bc1c7289bea